core/tracker/tracker.go

```diff
@@ -125,29 +125,73 @@ func (t *Tracker) Run(ctx context.Context) error {
 	}
 }
 
-// analyseDutyFailed detects if a duty failed. It returns true if the duty didn't complete the sigagg component.
-// If it failed, it also returns the component where it failed and a human friendly error message explaining why.
-func analyseDutyFailed(duty core.Duty, es []event) (bool, component, string) {
-	events := make([]event, len(es))
-	copy(events, es)
+// analyseDutyFailed detects if the given duty failed. It returns false if the duty didn't fail, i.e., the duty didn't get stuck and completes the sigAgg component.
+// It returns true if the duty failed. It also returns the component where the duty got stuck and a human friendly error message explaining why.
+func analyseDutyFailed(duty core.Duty, es map[core.Duty][]event) (bool, component, string) {
+	// dutyFailedComponent returns true if the duty failed. It also returns the component where the duty got stuck.  If the duty didn't get stuck, it
+	// returns the sigAgg component. It assumes that all the input events are for a single duty.
```

Comment:
Extract this as its own pure function, anonymous internal functions are only required when function state is updated